File: exercise/kotlin/day24/src/main/kotlin/usecases/ToyDeliveryUseCase.kt
package usecases

import arrow.core.Either
import arrow.core.flatMap
import domain.Toy
import domain.ToyRepository
import domain.core.Error
import domain.core.Error.Companion.anError

class ToyDeliveryUseCase(private val repository: ToyRepository) {
    fun handle(deliverToy: DeliverToy): Either<Error, Unit> =
        repository.findByName(deliverToy.desiredToy)
            .toEither { errorFor(deliverToy) }
            .flatMap { toy -> reduceStock(toy) }
            .map { }

    private fun reduceStock(toy: Toy): Either<Error, Toy> =
        toy.reduceStock()
            .tap { repository.save(it) }
            // we save even in case of failure
            .tapLeft { repository.save(toy) }
            .map { toy }

    private fun errorFor(deliverToy: DeliverToy): Error =
        anError("Oops we have a problem... we have not built the toy: ${deliverToy.desiredToy}")
}

File: exercise/scala/day24/src/main/scala/usecases/ToyDeliveryUseCase.scala
package usecases

import domain.core.Error
import domain.core.Error.anError
import domain.{Toy, ToyRepository}
import domain.core.Extensions.let

import scala.util.Either

class ToyDeliveryUseCase(private val repository: ToyRepository) {
  def handle(deliverToy: DeliverToy): Either[Error, Unit] =
    repository.findByName(deliverToy.desiredToy)
      .toRight(errorFor(deliverToy))
      .flatMap(toy => reduceStock(toy))
      .map(_ => ())

  private def reduceStock(toy: Toy): Either[Error, Toy] =
    toy.reduceStock()
      .let(_ => repository.save(toy))
      .map(_ => toy)

  private def errorFor(deliverToy: DeliverToy): Error =
    anError(s"Oops we have a problem... we have not built the toy: ${deliverToy.desiredToy}")
}